src/Features/CSharp/Portable/EmbeddedLanguages/CSharpTestEmbeddedLanguageClassifier.cs

```diff
@@ -0,0 +1,257 @@
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CSharpTestEmbeddedLanguageClassifier()
+        {
+        }
+
+        private static TextSpan FromBounds(VirtualChar vc1, VirtualChar vc2)
+            => TextSpan.FromBounds(vc1.Span.Start, vc2.Span.End);
+
+        public void RegisterClassifications(EmbeddedLanguageClassificationContext context)
+        {
+            var cancellationToken = context.CancellationToken;
+
+            var token = context.SyntaxToken;
+            var semanticModel = context.SemanticModel;
+            var compilation = semanticModel.Compilation;
+
+            if (token.Kind() is not (SyntaxKind.StringLiteralToken or SyntaxKind.SingleLineRawStringLiteralToken or SyntaxKind.MultiLineRawStringLiteralToken))
+                return;
+
+            var virtualCharsWithMarkup = CSharpVirtualCharService.Instance.TryConvertToVirtualChars(token);
+            if (virtualCharsWithMarkup.IsDefaultOrEmpty)
+                return;
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            // Simpler to only support literals where all characters/escapes map to a single utf16 character.  That way
+            // we can build a source-text as a trivial O(1) view over the virtual char sequence.
+            if (virtualCharsWithMarkup.Any(static vc => vc.Utf16SequenceLength != 1))
+                return;
+
+            using var _ = ArrayBuilder<TextSpan>.GetInstance(out var markdownSpans);
+            var virtualCharsWithoutMarkup = StripMarkupCharacters(virtualCharsWithMarkup, markdownSpans);
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            foreach (var span in markdownSpans)
+                context.AddClassification(ClassificationTypeNames.TestCodeMarkdown, span);
+
+            var encoding = semanticModel.SyntaxTree.Encoding;
+            var testFileSourceText = new VirtualCharSequenceSourceText(virtualCharsWithoutMarkup, encoding);
+
+            var testFileTree = SyntaxFactory.ParseSyntaxTree(testFileSourceText, semanticModel.SyntaxTree.Options, cancellationToken: cancellationToken);
+            var compilationWithTestFile = compilation.RemoveAllSyntaxTrees().AddSyntaxTrees(testFileTree);
+            var semanticModeWithTestFile = compilationWithTestFile.GetSemanticModel(testFileTree);
+
+            var start = virtualCharsWithoutMarkup[0].Span.Start;
+            context.AddClassification(
+                ClassificationTypeNames.TestCode,
+                TextSpan.FromBounds(
+                    start,
+                    virtualCharsWithoutMarkup.Last().Span.End));
+
+            var testFileClassifiedSpans = Classifier.GetClassifiedSpans(
+                context.SolutionServices,
+                project: null,
+                semanticModeWithTestFile,
+                new TextSpan(0, virtualCharsWithoutMarkup.Length),
+                ClassificationOptions.Default,
+                cancellationToken);
```

Comment:
will add comment here.  but it's ok to pass `project: null` here as we're not doing any nested embedded classification within this string.